### PR TITLE
Hitting Code Optimization

### PR DESCRIPTION
**Pull request by ImBeryl**
A complete rewrite of the TriggerColliderWeapon class to continue on the necessary code refactor the old game needed.

- Replaced multiple GetComponent calls with a few TryGetComponents.
- Stored components & GameObjects as variables instead of finding them every time the code was called.
- In general, simplification of the code that should both optimize the game and allow to easily implement new features in the future.

### CODE REVIEW

File: Assets/Scripts/Characters/AHSSShotGunCollider.cs [hunk 1]
```diff
@@ -140,7 +140,7 @@ private void OnTriggerStay(Collider other)
                         {
                             GameObject.Find("MainCamera").GetComponent<IN_GAME_MAIN_CAMERA>().startSnapShot2(item.transform.position, num8, null, 0.02f);
                         }
-                        item.transform.root.GetComponent<ColossalTitan>().titanGetHit(base.transform.root.gameObject.GetPhotonView().viewID, num8);
+                        //item.transform.root.GetComponent<ColossalTitan>().titanGetHit(base.transform.root.gameObject.GetPhotonView().viewID, num8);
```
Comment (Lithrun) [inline Assets/Scripts/Characters/AHSSShotGunCollider.cs]: Why is this commented out? This seems like that the colossal titan can no longer be harmed by AHSS now, which isn't something we want. Should revert this comment

File: Assets/Scripts/Characters/Titan/Body/BodyPart.cs [hunk 2]
```diff
@@ -5,6 +5,7 @@ public enum BodyPart
         None,
         Eyes,
         Head,
+        Nape,
```
Comment (Lithrun) [inline Assets/Scripts/Characters/Titan/Body/BodyPart.cs]: This enum is never used

File: Assets/Scripts/Characters/Titan/ColossalTitan.cs [hunk 3]
```diff
@@ -445,14 +445,14 @@ private void Start()
                 base.photonView.RPC("setSize", PhotonTargets.AllBuffered, new object[] { this.size });
                 this.lagMax = 150f + (this.size * 3f);
                 this.healthTime = 0f;
-                this.maxHealth = this.NapeArmor;
+                this.maxHealth = this.Health;
```
Comment (Lithrun) [inline Assets/Scripts/Characters/Titan/ColossalTitan.cs]: Avoid using "this" in newly added code

File: Assets/Scripts/Characters/Human/TriggerColliderWeapon.cs [hunk 4]
```diff
@@ -9,360 +9,238 @@ public class TriggerColliderWeapon : MonoBehaviour
...
         {
             Transform transform = titan.transform.Find("BodyPivot/HeadPos");
-            Vector3 to = base.transform.position - transform.transform.position;
+            Vector3 to = transform.position - transform.transform.position;
             return (Vector3.Angle(-transform.transform.forward, to) < 70f);
         }
         return false;
     }
 
     public void DummyNapeHit(DummyTitan titan)
```
Comment (Lithrun) [inline Assets/Scripts/Characters/Human/TriggerColliderWeapon.cs:24]: This method is unused. I think this should be somewhere within the case "titanneck"near line 109

File: Assets/Scripts/Characters/Human/TriggerColliderWeapon.cs [hunk 5]
```diff
@@ -9,360 +9,238 @@ public class TriggerColliderWeapon : MonoBehaviour
...
+        titan.photonView.RPC(nameof(DummyTitan.GetHit), PhotonTargets.All, new object[] { damage });
+        manager.netShowDamage(damage);
     }
 
-    public void clearHits()
+    public void ClearHits()
     {
-        this.currentHitsII = new ArrayList();
-        this.currentHits = new ArrayList();
+        currentHitsII = new ArrayList();
```
Comment (Lithrun) [inline Assets/Scripts/Characters/Human/TriggerColliderWeapon.cs]: Rather use .Clear() than using creating a new ArrayList

**Review by Lithrun — CHANGES_REQUESTED** [review]
100 lines less, that feels good! A few improvements can be made, so see my comments

File: Assets/Scripts/Characters/Human/TriggerColliderWeapon.cs (hunk 4 above)
Comment (ImBeryl) [reply Assets/Scripts/Characters/Human/TriggerColliderWeapon.cs:24]: Ah, true, thought it was used somewhere else, should be fixed in the next commit.

File: Assets/Scripts/Characters/Human/TriggerColliderWeapon.cs (hunk 5 above)
Comment (ImBeryl) [reply Assets/Scripts/Characters/Human/TriggerColliderWeapon.cs]: Though the difference might be minimal, I remember reading that `new ArrayList()` performed better vs `.Clear()`. Fixed in the latest commit.

File: Assets/Scripts/Characters/AHSSShotGunCollider.cs (hunk 1 above)
Comment (ImBeryl) [reply Assets/Scripts/Characters/AHSSShotGunCollider.cs]: Oh, commented it out when I was testing the changes, replaced it with the newest hit method.

File: Assets/Scripts/Characters/Human/TriggerColliderWeapon.cs [hunk 6]
```diff
@@ -109,8 +92,13 @@ private void OnTriggerStay(Collider collider)
             case "titanneck":
                 if (collider.gameObject.TryGetComponent(out HitBox item) && item.transform.root.TryGetComponent(out TitanBase titanBase))
                 {
-                    if(!IsBehind(item.transform.root.gameObject)) 
-                        return;
+                    if (titanBase.transform.Find("Amarture/Core/Controller_Body/hip/spine/chest/neck/head") != null)
```
Comment (Lithrun) [inline Assets/Scripts/Characters/Human/TriggerColliderWeapon.cs]: titanBase has a property called "TitanBody" which contains a set of bodypart references. transform.Find should be avoided as it is a very expensive method. This goes for both line 95 and 97.

I'm not too sure if the Female and Colossal titan have this component added yet. If they do not, then please add the TitanBody and assign their head object to the TitanBody Head field. Then verify if the Colossal and Female titan can still be damaged.

**Review by Lithrun — CHANGES_REQUESTED** [review]
Final comment 👍

File: Assets/Scripts/Characters/Human/TriggerColliderWeapon.cs (hunk 6 above)
Comment (ImBeryl) [reply Assets/Scripts/Characters/Human/TriggerColliderWeapon.cs]: Didn't notice that one while moving code around, nice catch! Should be fixed in the most recent commit. (Tested normal, Colossal & Female titans, all were damagable)

**Review by Lithrun — APPROVED** [review]

**Pull request merged**